libpod/sqlite_state.go

```diff
@@ -0,0 +1,2535 @@
+			}
+		}
+	}()
+
+	for id, json := range ctrStates {
+		if _, err := tx.Exec("UPDATE TABLE ContainerState SET JSON=? WHERE ID=?;", json, id); err != nil {
+			return fmt.Errorf("updating container state: %w", err)
+		}
+	}
+	for id, json := range podStates {
+		if _, err := tx.Exec("UPDATE TABLE PodState SET JSON=? WHERE ID=?;", json, id); err != nil {
+			return fmt.Errorf("updating pod state: %w", err)
+		}
+	}
+	for name, json := range volumeStates {
+		if _, err := tx.Exec("UPDATE TABLE VolumeState SET JSON=? WHERE Name=?;", json, name); err != nil {
+			return fmt.Errorf("updating volume state: %w", err)
+		}
+	}
+
+	if _, err := tx.Exec("DELETE FROM ContainerExitCode;"); err != nil {
+		return fmt.Errorf("removing container exit codes: %w", err)
+	}
+
+	if _, err := tx.Exec("DELETE FROM ContainerExecSession;"); err != nil {
+		return fmt.Errorf("removing container exec sessions: %w", err)
+	}
+
+	if err := tx.Commit(); err != nil {
+		return fmt.Errorf("committing transaction: %w", err)
+	}
+
+	return nil
+}
+
+// GetDBConfig retrieves runtime configuration fields that were created when
+// the database was first initialized
+func (s *SQLiteState) GetDBConfig() (*DBConfig, error) {
+	if !s.valid {
+		return nil, define.ErrDBClosed
+	}
+
+	cfg := new(DBConfig)
+	var staticDir, tmpDir, graphRoot, runRoot, graphDriver, volumeDir string
+
+	row := s.conn.QueryRow("SELECT StaticDir, TmpDir, GraphRoot, RunRoot, GraphDriver, VolumeDir FROM DBConfig;")
+
+	if err := row.Scan(&staticDir, &tmpDir, &graphRoot, &runRoot, &graphDriver, &volumeDir); err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return cfg, nil
+		}
+		return nil, fmt.Errorf("retrieving DB config: %w", err)
+	}
+
+	cfg.LibpodRoot = staticDir
+	cfg.LibpodTmp = tmpDir
+	cfg.StorageRoot = graphRoot
+	cfg.StorageTmp = runRoot
+	cfg.GraphDriver = graphDriver
+	cfg.VolumePath = volumeDir
+
+	return cfg, nil
+}
+
+// ValidateDBConfig validates paths in the given runtime against the database
+func (s *SQLiteState) ValidateDBConfig(runtime *Runtime) (defErr error) {
+	if !s.valid {
+		return define.ErrDBClosed
+	}
+
+	storeOpts, err := storage.DefaultStoreOptions(rootless.IsRootless(), rootless.GetRootlessUID())
+	if err != nil {
+		return err
+	}
+
+	const createRow = `
+        INSERT INTO DBconfig VALUES (
+                ?, ?, ?,
+                ?, ?, ?,
+                ?, ?, ?
+        );`
+
+	var (
+		os, staticDir, tmpDir, graphRoot, runRoot, graphDriver, volumePath string
+		runtimeOS                                                          = goruntime.GOOS
+		runtimeStaticDir                                                   = filepath.Clean(s.runtime.config.Engine.StaticDir)
+		runtimeTmpDir                                                      = filepath.Clean(s.runtime.config.Engine.TmpDir)
+		runtimeGraphRoot                                                   = filepath.Clean(s.runtime.StorageConfig().GraphRoot)
```

Comment:
long-term these could be ripe for a struct, esp bc you do a bunch of validation afterwards, take it or leave it